# Yeast breeding experiment

In this example, we want to model a cross of two diploid yeast strains. Depending on which sequence data you have
available, you can take multiple approaches depending on which genome data you have available. In this document we will
demonstrate a workflow that makes us of variant call files, and a workflow that uses whole genome alignment. In both
cases we study two beer yeasts from the [1002 Yeast Genome collection](http://1002genomes.u-strasbg.fr/): Orval trappist
yeast strain DBVPG6695 from Belgium and American ale strain 1.3_Safale_US05 (codenames BRQ and CFD).

First we set up a new gen repository to store and track our sequences.

```console
gen init
```

## Starting from variant calls

The dataset we are working with includes variant call files that were generated by aligning sequencing reads to the
S288C reference genome, specifically version R64-1-1. We'll start by downloading this reference sequence from the
Saccharomyces Genome Database (SGD) and extracting the archive.

```console
wget http://sgd-archive.yeastgenome.org/sequence/S288C_reference/genome_releases/S288C_reference_genome_R64-1-1_20110203.tgz
tar -xzvf S288C_reference_genome_R64-1-1_20110203.tgz 
```

Let's take a look at the contigs present in the main fasta file:

```console
cd S288C_reference_genome_R64-1-1_20110203/
grep ">" S288C_reference_sequence_R64-1-1_20110203.fsa
```

This results in the following output, showing that each contig is named after its  NCBI identifier (e.g. NC_001133):

```
>ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]
>ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]
>ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]
>ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]
>ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]
>ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]
>ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]
>ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]
>ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]
>ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]
>ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]
>ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]
>ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]
>ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]
>ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]
>ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]
>ref|NC_001224| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [location=mitochondrion] [top=circular]
```

Next, we download the variant calls. These are stored in a large Genomic VCF (GVCF) file that includes variants from all samples in
the dataset. We download a compressed version of this file from the the 1002 Yeast Genomes project and also take a look
at the contigs mentioned in the header portion of the file:

```console
wget http://1002genomes.u-strasbg.fr/files/1011Matrix.gvcf.gz
gzip -cd 1011Matrix.gvcf.gz | head -n 100 | grep contig
```

From the output we can see that unfortunately the variant call file refers to contigs not by their NCBI identifier, but by the chromosome number:

```
##contig=<ID=chromosome1,length=230218>
##contig=<ID=chromosome2,length=813184>
##contig=<ID=chromosome3,length=316620>
##contig=<ID=chromosome4,length=1531933>
##contig=<ID=chromosome5,length=576874>
##contig=<ID=chromosome6,length=270161>
##contig=<ID=chromosome7,length=1090940>
##contig=<ID=chromosome8,length=562643>
##contig=<ID=chromosome9,length=439888>
##contig=<ID=chromosome10,length=745751>
##contig=<ID=chromosome11,length=666816>
##contig=<ID=chromosome12,length=1078177>
##contig=<ID=chromosome13,length=924431>
##contig=<ID=chromosome14,length=784333>
##contig=<ID=chromosome15,length=1091291>
##contig=<ID=chromosome16,length=948066>
```

Because the order of the contigs is the same in both files, we can use [seqtk](https://github.com/lh3/seqtk) to rename
the contigs in the reference fasta file to `chromosome` followed by an index number.

```console
seqtk rename S288C_reference_sequence_R64-1-1_20110203.fsa chromosome > S288C_reference_renamed.fa

Now we can import the refernce sequence into to our gen repository:

```console
gen import --fasta "S288C_reference_renamed.fa" 
```

Next, we use bcftools to extract the variant calls for the two samples we are interested in (namely BRQ and CFD). This takes about 15 minutes to complete on a
laptop.

```console
bcftools view -s BRQ -o BRQ.vcf 1011Matrix.gvcf.gz 
bcftools view -s CFD -o CFD.vcf 1011Matrix.gvcf.gz 
```

We can now load the individual VCF files into gen using an `update` command. Gen uses the chromosome/contig and sample
names in the VCF file to find the appropriate graphs to update or create. In our case we want to combine the variants,
so we tell gen to create a sample called F1.

(BUG: trying to specify a sample when the VCF already has samples should error out, or redirect the variants (if the
count matches), see bug report in Notion)

```console
gen update --vcf BRQ_2.vcf --sample F1
gen update --vcf CFD_2.vcf --sample F1
```

Workaround:
```
bcftools reheader -s sample_mapping.txt -o BRQ_3.vcf BRQ_2.vcf
bcftools reheader -s sample_mapping.txt -o CFD_3.vcf CFD_2.vcf
```

If we want to study a specific region in detail, we can extract it using the `derive subgraph` command:

```console

```

Lastly, we export the F1 sample to a GFA file that can be used by graph-aware sequence aligners.

```console
gen export --gfa cross.gfa
```

## Starting from assembled genomes

First we download the genome assembly archive and extract the genome sequence for the strains we want to mate.

```
wget http://1002genomes.u-strasbg.fr/files/1011Assemblies.tar.gz
tar -xzf 1011Assemblies.tar.gz GENOMES_ASSEMBLED/BRQ_4.re.fa GENOMES_ASSEMBLED/CFD_4.re.fa
```

Then we use [Cactus](https://doi.org/10.1038/s41586-020-2871-y) to create a graph via whole genome alignment. The
Comparative Genomics Toolkit provides a Docker container that includes all the infrastructure needed to do so. We launch
the container and bind our working directory to the /data directory in the container:

```console
docker run --mount type=bind,source=$PWD,target=/data -it quay.io/comparative-genomics-toolkit/cactus:v2.9.0 bash
```

We tell Cactus which genomes to align by providing a text file containing names and file paths. You can set this up
manually, or generate it for all fasta files in our current directory and subdirectories:

```console
find . -type f -name "*.fa" -printf "%f %p\n" | awk '{ sub(/\..*$/, "", $1); print $1, $2 }' > file_list.txt
```

Then we can start the Cactus pipeline using the command list below. `./js` specifies where intermediate files are
stored, outName and outDir influence the pipeline, and `--gfa` tells Cactus to output a GFA file. While this is a graph
based alignment, we must still designate one genome as reference, in this case we pick BRQ.

```console
cactus-pangenome ./js file_list.txt --outDir ./cactus_output --outName cross --reference BRQ_4 --gfa
```

After this completes (approximately 10 minutes), we can load it back into gen by running:

```console
gen import --gfa cactus_output/cross.gfa
```

## References

1. https://www.nature.com/articles/s41586-018-0030-5
